### ISSUE TITLE
Add the ability to import .exr with layer support as a TextureArray, add a resource similar to AtlasTexture that allows you to select a separate TextureArray layer

### ISSUE THREAD

**Issue by anunknowperson** [godotengine/godot-proposals#6302]
### Describe the project you are working on

2D game with heavy use of cutout/skeletal animation

### Describe the problem or limitation you are having in your project

Recently, I've been working on skeleton 2D animation, and I've found that transferring a layered picture from my graphics editor (SAI) to Godot is a time-consuming process that doesn't yield good results. For example, if my image has 50 moving parts, I would need to manually export 50 pictures to jpg and then set each picture to the right position in the editor. This process is not only time-consuming, but it can also create clutter in the file system of the project.
This same issue also applies to Cutout animation.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Support for importing exr as a TextureArray will allow you to transfer assembled layered images directly from the graphical editor to Godot without having to waste time on the same monotonous actions.

It will also greatly reduce the number of files in the file system of the project.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

1) The ability to import exr/psd either as a normal image or as a TextureArray.
2) A resource, similar in implementation to TextureAtlass or PSDLayer from my recent PR, which allows you to specify an image layer to use

### If this enhancement will not be used often, can it be worked around with a few lines of script?

I think this is basic functionality.

### Is there a reason why this should be core and not an add-on in the asset library?

Support for image abstraction with layers should be in the engine. 

.exr support is already in the engine, so it shouldn't be a problem to add layer support for it.

**Comment by anunknowperson** [comment]
Will superseed https://github.com/godotengine/godot/pull/73417

**Comment by anunknowperson** [comment]
Related:
https://github.com/godotengine/godot/issues/15050
https://github.com/godotengine/godot/issues/23723

**Comment by Calinou** [comment]
> For example, if my image has 50 moving parts, I would need to manually export 50 pictures to jpg and then set each picture to the right position in the editor. 

Is there no CLI interface or third-party tool to automate exporting each layer to a JPEG?

**Comment by fire** [comment]
@dsnopek 's idea of taking an exr and generating a 2d scene when the layers set properly might work for the blending and for usability.

**Comment by anunknowperson** [comment]
@fire 

The problem is that if the file contains many layers, the file system of the project will again become cluttered.

Maybe do something like Texture2DDictionary, similar to Texture2DArray, but with the ability to get a layer by name rather than by number?

**Comment by fire** [comment]
I was thinking like an entire PackedScene - one file for let's say 50 images. I can probably demo it to you using your psd importer.

**Comment by anunknowperson** [comment]
Does this mean that the images will be stored inside the scene file? 

I have a bad idea how this will work if I need to take these layers and add them to my character's scene and let's say move them around a bit. After that, if I make a change to the .psd file, would I have to redo everything again?

Also, if images are stored inside the scene does that mean they will be stored without compression?

**Comment by fire** [comment]
How would you like it be used? I wasn't able to find a good way.

**Comment by anunknowperson** [comment]
> How would you like it be used? I wasn't able to find a good way.

You mean in terms of implementation? I think something like TextureDictionary

And in terms of use in the editor, as I said for skeletal/cutout 2D animation.
https://docs.godotengine.org/en/stable/tutorials/animation/2d_skeletons.html

I would like to be able to use layers of the same PSD in Polygon instead of different images in this case. And that I don't have to assemble it all manually

**Comment by anunknowperson** [comment]
Also, in the .psd proposal someone said that for use in visual novels for example to change facial expressions, you can make a resource like AtlassTexture or PSDLayer and just change the name of the layer, not hide/show nodes.

**Comment by Calinou** [comment]
> The problem is that if the file contains many layers, the file system of the project will again become cluttered.

We can create a folder that matches the PSD's filename (minus the extension), then write all textures within.

**Comment by anunknowperson** [comment]
>We can automatically create a folder that matches the PSD's filename (minus the extension), then write all textures inside.

I think it's a good option.

**Comment by fire** [comment]
I'll summarize the three modes described earlier in the thread:

## One image

The entire pdf is one Godot Engine image.

## Extract textures

We were recommended against creating a folder for .glb.

Implement a mode like the glb where the psd is extracted in a file name concatenated with the layer name scheme.

## Convert to PackedScene

> I would like to be able to use layers of the same PSD in Polygon instead of different images in this case.

If it's in a packed scene we can hide or modify the layers which aren't forced to document size.

**Comment by Calinou** [comment]
> Having folders is a problem, where we can resolve by concatenating the filename and the layer.

If you have a lot of PSDs in the same folder, you will quickly get a single folder with thousands of images :slightly_smiling_face:

**Comment by dsnopek** [comment]
> Modify Texture2DArray to have both names and index

I don't think we should make changes to `Texture2DArray` for this. `Texture2DArray` directly relates to a specific resource on the GPU: `GL_TEXTURE_2D_ARRAY` in OpenGL, and... something in Vulkan (I don't know Vulkan at all :-)). So, the .psd shouldn't become a `Texture2DArray` unless the developer intends to pass it to the GPU as a 2D texture array.

**Comment by Calinou** [comment]
> I don't think we should make changes to `Texture2DArray` for this. `Texture2DArray` directly relates to a specific resource on the GPU: `GL_TEXTURE_2D_ARRAY` in OpenGL, and... something in Vulkan (I don't know Vulkan at all :-)). So, the .exr shouldn't become a `Texture2DArray` unless the developer intends to pass it to the GPU as a 2D texture array.

Indeed, as this would require modifying the core 2D3D/ shaders to allow using a Texture2DArray anywhere a Texture2D can be used. This would make the shaders more complex and require more shader versions.

**Comment by YYF233333** [comment]
Just a noob idea, is it possible to impl a custom resource for such layered images? Like a PSDTexture, with hide_layer() and  to_texture2d(). Then reading the psd at runtime, merge layers and dump a texture2d. No extra file generated, as compact as the source file, and if the code not fit in the engine, we could make a plugin(? not sure). Don't know if this approach meets the requirements of skeleton animation.

**Comment by anunknowperson** [comment]
> Just a noob idea, is it possible to impl a custom resource for such layered images?

That's kind of my original idea.

Create an abstract resource like TextureArray for images stored by layer name.

And then add exr import to the editor and .psd import to GDExtension, for example.

And I haven't really figured out why that's a bad option yet.

**Comment by anunknowperson** [comment]
@fire @Calinou 
I apologize for asking again, but could you clarify if creating a new resource that stores images is a viable option? Or for some reason we need to implement it in another way?

**Comment by dsnopek** [comment]
@anunknowperson:

> I apologize for asking again, but could you clarify if creating a new resource that stores layers by names is a viable option? Or for some reason we need to implement it in another way?

In my opinion, it's fine, so long as the way it fits into the editor UI is reasonably easy to use. I'd love to hear the "stories" of how a user would take a multi-layer .exr file, and then step-by-step, how they'd use it for various tasks.

For example, what steps does the user do in order to:

- Create a number of `Sprite2D`'s using different layers for each
- Create an `AnimatedSprite2D` where the `SpriteFrames` are composed of different layers
- Do the same as the last point but with a `Sprite2D` and an `AnimationPlayer`
- Create a `TileSet` where the tiles come from various layers

I personally can't yet picture how this new resource type would fit into the UI for all the different cases where we currently use a `Texture2D`.

**Comment by YYF233333** [comment]
> @anunknowperson:
> In my opinion, it's fine, so long as the way it fits into the editor UI is reasonably easy to use. I'd love to hear the "stories" of how a user would take a multi-layer .exr file, and then step-by-step, how they'd use it for various tasks.

I'd like to share an example.
<img src="https://user-images.githubusercontent.com/67693830/220165492-4641f785-0fc3-465d-8965-5a2f3e781dea.png" width="40%" height="40%">
This is one of my character image, it has a body layer, a clothes layer and 13 face layers. With a layer Texture I can keep all these layers in memory. When I want to change the facial expression, I choose one face, hide others, combine it with body and clothes, then dump them into a Texture2D and show with a Sprite.

In conclusion, layer texture can be used when you have an image with replacable parts. It is not convenient if you import then as texture2d because they are simple just not expected to show alone.

**Comment by dsnopek** [comment]
> I'd like to share an example.
>
> This is one of my character image, it has a body layer, a clothes layer and 13 face layers. With a layer Texture I can keep all these layers in memory. When I want to change the facial expression, I choose one face, hide others, combine it with body and clothes, then dump them into a Texture2D and show with a Sprite.
>
> In conclusion, layer texture can be used when you have an image with replacable parts. It is not convenient if you import then as texture2d because they are simple just not expected to show alone.
>
> I think there is no need to have the ability to change the texture in editor, just tell me it is a psd/exr, I will open them in other software. (a check box to hide certian layer is appreciated though)

Thanks! However, I still can't really picture exactly how you'd use the Godot editor in this case.

Does it look like this:

1. Drop the .exr file with all the layers into your Godot project
2. Right click in the file system dock, click "New" -> "Resource..."
3. Select a `Texture2DFromLayer` resource (I'm imagining that this would be created as part of the PR to add this feature)
5. Save it as "my_texture.tres"
6. Drag the .exr file from the file system dock into a "Source" field on the inspector for the resource
7. In some kind of specialized editor (maybe in the bottom dock where we already have "Animation", "Shader Editor", etc), select the layers you wish to hide and show to make the character with the right facial expression
8. Create a `Sprite2D` node, and drag the "my_texture.tres" into the "Texture" field of the node

Or, if the the step-by-step "story" looks different than that, I'd really love to hear it!

I could also imagine importing the .exr file as a `PackedScene`, and then you'd make a "New Inherited Scene", where you can hide/show the different layers? But @anunknowperson above was specifically talking about making some new resource types to refer to layers, which is why my "story" above goes that route.

I'm just trying to get a handle on what exactly we are proposing from the perspective of a user of the Godot editor.

**Comment by YYF233333** [comment]
I want to do this:
```GDScript
extends Sprite2D

func _ready():
    var img: LayerTexture = load("path/to/file")
    img.hide_layer(x)
    self.texture = img.to_texture2d()
```

> I think there is no need to have the ability to change the texture in editor, just tell me it is a psd/exr, I will open them in other software. (a check box to hide certian layer is appreciated though)

I means some checkbox in the right dock, something like:
- [ ] layer name1
- [ ] layer name2
- [x] layer name3

tick is show, untick is hide. It is just a qol feature.

seems this texture is unlikely to appear in the editor. (all of them will be convert to texture2d before assign) Sorry my mind is kind of not that clear, let's just ignore the checkbox part :)

> I could also imagine importing the .exr file as a `PackedScene` with a `Sprite2D` for each layer, and then you'd make a "New Inherited Scene", where you can hide/show the different layers?

That's exactly the workaround now, the cons is that we will have tons of png in the filesystem, and we cannot do hot reload.

**Comment by dsnopek** [comment]
> I want to do this:
> ```GDScript
> extends Sprite2D
> 
> func _ready():
>     var img: LayerTexture = load("path/to/file")
>     img.hide_layer(x)
>     self.texture = img.to_texture2d()
> ```

Ah, so for you, this is something you want to use via GDScript?

If we're implementing this feature for artists, it would make sense to me to focus on making it easy-to-use via the editor, without requiring artists to write code.

**Comment by YYF233333** [comment]
> If we're implementing this feature for artists, it would make sense to me to focus on making it easy-to-use via the editor, without requiring artists to write code.

I think there are some misunderstanding, artists don't write code, programmers write them.

What I want to illustrate by this piece of code is the API I expose to script writer. They simple write lines in txt like `show character a with emotion happy` and programmers take the responsibility to implement this effect.

Writers don't know that a character has 10 layers, they just think we have different images like `a_happy.png`, `a_sad.png`, exposing this detail to them will make things more complex.

You may ask why I implement this in such complex way, well, reasons:
- I get images in psd format, things get worse for small teams (like my team), large teams may feel it _hard_ to change the workflow, but for small teams it is _impossible_ because images are usually outsourcing, I cannot ask artists to change their workflow.
- export all combinations to png will bloat game size, even webp doesn't help because you are saving mostly same things in different images.

Also you would find accessing this texture in editor meaningless. I cannot control what's in the script, I interpret it at runtime so I have to prepare for all possibilities.

Anyway, I cannot imagine people reaching this feature without writing even one line of code. By saying "artists" I assume you means people who are good at art and code at the same time, not people whose computer are just "Photoshop Launcher".(not meant to be offensive) If someone really have difficulty in coding, programmers in their teams are willing to help :)

p.s. I propose this solution because it has least impact to the editor (just a new resource), feel free to point out if I make any mistake :)

**Comment by YYF233333** [comment]
This picture may help to explain the idea more clearly:
![struct](https://user-images.githubusercontent.com/67693830/220258333-c09a372d-b1bb-4c67-a609-dfb1867a2dc3.jpg)

**Comment by fire** [comment]
```
extends Sprite2D

func _ready():
    var img: LayerTexture = load("path/to/file")
    img.hide_layer(x)
    self.texture = img.to_texture2d()
```

I am interpreting the script will do this:

1. Load psd / exr to a resource
3. At this point the texture will have all the layers in memory copied for example from the psd or the exr.
4. Hide any amount of layers.
5. Take this resource and bake a different texture 2d at runtime which may or may not be slow.

This is compared to:

1. Load the psd / exr into a packed scene. This is inspect-able and artist editable
2. The programmer will now have a tree of texture 2ds in the scene
3. The programmer can now extract the images from the nodes
4. Then apply the image blend rect method to get one texture 2d. https://docs.godotengine.org/en/latest/classes/class_image.html#class-image-method-blend-rect
5. Use the texture2d for future use.

**Comment by anunknowperson** [comment]
> Load the psd / exr into a packed scene. This is inspect-able and artist editable

The problem with the generated scenes approach is that if I need to do some manipulations with the images in the engine, for example set the polygons and bone weights on Polygon2D, then I need to transfer the sprites from the generated scene into my scene. Then if I change something in .exr file, I will need to transfer objects individually to my scene again. Also it turns out that the images will be stored within my scene and not in the resource.

**Comment by anunknowperson** [comment]
I think now to make an example of how I want to implement it and show here.

**Comment by YYF233333** [comment]
> This is compared to:
> 
> 1. Load the psd / exr into a packed scene. This is inspect-able and artist editable
> 2. The programmer will now have a tree of texture 2ds in the scene
> 3. The programmer can now extract the images from the nodes
> 4. Then apply the image blend rect method to get one texture 2d. https://docs.godotengine.org/en/latest/classes/class_image.html#class-image-method-blend-rect
> 5. Use the texture2d for future use.


> ~That's exactly the workaround now,~ Workaround I used now is split psd by layer, reconstruct them in godot as a scene. The cons is that we will have tons of png in the filesystem, and we cannot do hot reload. If this process can be automated it is as fine as having a new texture :)

The key point is automation, which needs some modify in editor, this approach is fine for me mainly.

I didn't realize we can bake multiple texture into one, thanks!

**Comment by dsnopek** [comment]
@YYF233333:

> Anyway, I cannot imagine people reaching this feature without writing even one line of code. By saying "artists" I assume you means people who are good at art and code at the same time, not people whose computer are just "Photoshop Launcher".(not meant to be offensive) If someone really have difficulty in coding, programmers in their teams are willing to help :)

The "a programmer will integrate the art into the game" approach is certainly a valid workflow that's used on many teams. I've worked on projects like this too! :-)

However, on some teams, the artists load their art into the editor themselves. This has the advantage of empowering artists (ex. they can iterate on their art quickly by seeing what it looks like in the context of the game, without needing to wait for a programmer), and can also allow the artists to make the animations in engine as well, which for the most part can also be done without programming. I've also worked on projects that used this sort of workflow, and using the Godot editor isn't usually a problem for artists - getting them up-to-speed on version control is usually the biggest hurdle. :-)

In any case, Godot is an engine with an editor (there's certainly many game engines out there where everything MUST be integrated in code), and so I think we should provide a reasonable way to use this feature in the editor without _requiring_ code (certainly you can always use code for more advanced use cases)

**Comment by YYF233333** [comment]
> I've also worked on projects that used this sort of workflow, and using the Godot editor isn't usually a problem for artists

Oh, if that's the case we do need this feature in the editor. Sadly I have no idea what it should be like either. :-( Maybe @anunknowperson has this need.

**Comment by anunknowperson** [comment]
Unfortunately I don't have enough free time at the moment, but I'll get back to it in a few months if someone else doesn't take the trouble to implement it.
